File: civ/src/embedding.ts
// Embeddings simplificados para o MVP local (sem chamada a um provedor de
// embeddings externo): hashing trick (FNV-1a) sobre bag-of-words, projetado
// em 64 dimensões e normalizado em L2. Não é semanticamente tão rico quanto
// um embedding de modelo de linguagem, mas é determinístico, roda 100%
// offline e é suficiente para a recuperação por similaridade do RAG nesta
// demonstração. A arquitetura de produção (Seção 4 da documentação técnica)
// usa embeddings do provedor do LLM. A MESMA função (mesmo hash, mesma
// dimensão) precisa existir no Orquestrador Python — ver orchestrator/embedding.py.

const DIM = 64;

function fnv1a(str: string): number {
  let hash = 0x811c9dc5;
  for (let i = 0; i < str.length; i++) {
    hash ^= str.charCodeAt(i);
    hash = Math.imul(hash, 0x01000193);
  }
  return hash >>> 0;
}

const STOPWORDS = new Set(["de", "a", "o", "que", "e", "do", "da", "em", "um", "uma", "para", "com", "no", "na", "os", "as", "por", "seu", "sua", "ou", "se", "meu", "minha"]);

export function embed(text: string): number[] {
  const vec = new Array(DIM).fill(0);
  const tokens = text
    .toLowerCase()
    .normalize("NFD").replace(/[̀-ͯ]/g, "") // remove acentos
    .split(/[^a-z0-9]+/)
    .filter((t) => t.length > 1 && !STOPWORDS.has(t));

  for (const token of tokens) {
    const idx = fnv1a(token) % DIM;
    vec[idx] += 1;
  }
  const norm = Math.sqrt(vec.reduce((acc, v) => acc + v * v, 0)) || 1;
  return vec.map((v) => v / norm);
}

File: orchestrator/embedding.py
"""
embedding.py — porta em Python do civ/src/embedding.ts.

Implementa EXATAMENTE o mesmo algoritmo (hashing trick FNV-1a sobre
bag-of-words, 64 dimensões, normalização L2) para que os vetores
calculados aqui sejam comparáveis (via distância pgvector `<->`) com
os vetores já gravados em knowledge_base.embedding por civ/src/seed.ts.

MVP: isto substitui um provedor de embeddings real (que exigiria uma
chave de API paga). A arquitetura de produção documentada usa
embeddings do provedor do LLM (Seção 4 da documentação técnica); aqui
usamos um esquema determinístico e local para a demonstração poder
rodar 100% offline, com a MESMA função implementada nos dois lados
(TypeScript e Python) — ver civ/src/embedding.ts.
"""
import math
import re
import unicodedata
from typing import List

DIM = 64
FNV_OFFSET_BASIS = 0x811C9DC5
FNV_PRIME = 0x01000193
MASK32 = 0xFFFFFFFF

# Mesma lista de STOPWORDS de civ/src/embedding.ts
STOPWORDS = {
    "de", "a", "o", "que", "e", "do", "da", "em", "um", "uma", "para",
    "com", "no", "na", "os", "as", "por", "seu", "sua", "ou", "se",
    "meu", "minha",
}


def _fnv1a(token: str) -> int:
    h = FNV_OFFSET_BASIS
    for ch in token:
        h ^= ord(ch)
        h = (h * FNV_PRIME) & MASK32
    return h


def _strip_accents(text: str) -> str:
    nfd = unicodedata.normalize("NFD", text)
    return "".join(c for c in nfd if unicodedata.category(c) != "Mn")


def _tokenize(text: str) -> List[str]:
    lowered = _strip_accents(text.lower())
    raw = re.split(r"[^a-z0-9]+", lowered)
    return [t for t in raw if len(t) > 1 and t not in STOPWORDS]


def embed(text: str) -> List[float]:
    vec = [0.0] * DIM
    for token in _tokenize(text or ""):
        idx = _fnv1a(token) % DIM
        vec[idx] += 1
    norm = math.sqrt(sum(v * v for v in vec)) or 1
    return [v / norm for v in vec]


def to_pgvector_literal(vec: List[float]) -> str:
    return "[" + ",".join(f"{v:.8f}" for v in vec) + "]"
